icicle/appUtils/msm/msm.cu

```diff
@@ -880,109 +882,137 @@ void batched_bucket_method_msm(
 
-// the function computes msm using ssm
-template <typename S, typename P, typename A>
-void short_msm(S* h_scalars, A* h_points, unsigned size, P* h_final_result, cudaStream_t stream)
-{ // works up to 2^8
-  S* scalars;
-  A* a_points;
-  P* p_points;
-  P* results;
-
-  cudaMallocAsync(&scalars, sizeof(S) * size, stream);
-  cudaMallocAsync(&a_points, sizeof(A) * size, stream);
-  cudaMallocAsync(&p_points, sizeof(P) * size, stream);
-  cudaMallocAsync(&results, sizeof(P) * size, stream);
-
-  // copy inputs to device
-  cudaMemcpyAsync(scalars, h_scalars, sizeof(S) * size, cudaMemcpyHostToDevice, stream);
-  cudaMemcpyAsync(a_points, h_points, sizeof(A) * size, cudaMemcpyHostToDevice, stream);
-
-  // convert to projective representation and multiply each point by its scalar using single scalar multiplication
-  unsigned NUM_THREADS = size;
-  to_proj_kernel<<<1, NUM_THREADS, 0, stream>>>(a_points, p_points, size);
-  ssm_kernel<<<1, NUM_THREADS, 0, stream>>>(scalars, p_points, results, size);
-
-  P* final_result;
-  cudaMallocAsync(&final_result, sizeof(P), stream);
-
-  // assuming msm size is a power of 2
-  // sum all the ssm results
-  NUM_THREADS = size;
-  sum_reduction_kernel<<<1, NUM_THREADS, 0, stream>>>(results, final_result);
-
-  // copy result to host
-  cudaStreamSynchronize(stream);
-  cudaMemcpyAsync(h_final_result, final_result, sizeof(P), cudaMemcpyDeviceToHost, stream);
-
-  // free memory
-  cudaFreeAsync(scalars, stream);
-  cudaFreeAsync(a_points, stream);
-  cudaFreeAsync(p_points, stream);
-  cudaFreeAsync(results, stream);
-  cudaFreeAsync(final_result, stream);
+template <typename S, typename A, typename P>
+void msm(S* scalars, A* points, unsigned size, P* result)
+{
+  MSMConfig config = {
+    false,     // scalars_on_device
+    true,      // scalars_montgomery_form
+    size,      // points_size
+    1,         // precompute_factor
+    false,     // points_on_device
+    true,      // points_montgomery_form
+    1,         // batch_size
+    false,     // result_on_device
+    16,        // c
+    S::NBITS,  // bitsize
+    false,     // big_triangle
+    10,        // large_bucket_factor
+    0,         // device_id
+    0          // stream
+  };
+  msm_internal(scalars, points, size, config, result);
 }
 
-// the function computes msm on the host using the naive method
-template <typename A, typename S, typename P>
-void reference_msm(S* scalars, A* a_points, unsigned size)
+/**
+ * Extern version of [msm_internal](@ref msm_internal) function with the following values of template parameters 
+ * (where the curve is given by `-DCURVE` env variable during build):
+ *  - `S` is the [scalar field](@ref scalar_t) of the curve;
+ *  - `A` is the [affine representation](@ref affine_t) of curve points;
+ *  - `P` is the [projective representation](@ref projective_t) of curve points.
+ * @returns 0 if no error occured and -1 otherwise.
+ */
+extern "C" int msm_internal_cuda(
+  curve_config::scalar_t scalars[],
```

Comment:
isn't curve_config::scalar_t scalars* more common in our code?

Reply:
tbh I just copied the code that we had before. can rename for consistency, there aren't significant differences between the two though, right?